File: clojure/phone-number/src/phone_number.clj
(ns phone-number
  (:require [clojure.string :as string]))

(defn- parse [num-string]
  (let [cleaned-num-string (string/replace num-string #"\D+" "")
        match (re-matches #"1?([2-9]\d{2})([2-9]\d{2})(\d{4})" cleaned-num-string)
        [area-code exchange-code subscriber-number] (rest match)]
    (sorted-map
      :area-code (or area-code "000")
      :exchange-code (or exchange-code "000")
      :subscriber-number (or subscriber-number "0000"))))

(defn number [num-string]
  (string/join (vals (parse num-string))))

(defn area-code [num-string]
  (:area-code (parse num-string)))

(defn pretty-print [num-string]
  (let [parsed (parse num-string)]
    (str "(" (:area-code parsed) ") " (:exchange-code parsed) "-" (:subscriber-number parsed))))

File: javascript/phone-number/phone-number.js
export class PhoneNumber {
    constructor(input) {
        const cleanedInput = input.replace(/[^0-9]/g, '');
        const match = /^1?([2-9][0-9]{2}[2-9][0-9]{6})$/.exec(cleanedInput);
        this.parsedNumber = match ? match[1] : null;
    }

    number() {
        return this.parsedNumber;
    }
}
